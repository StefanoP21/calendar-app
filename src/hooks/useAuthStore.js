import { useDispatch, useSelector } from 'react-redux';
import { calendarApi } from '../api';
import { clearError, onChecking, onLogin, onLogout } from '../context';

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth', { email, password });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch ({ response }) {
      let errorMessage = 'Error en el servidor';

      if (response.data.msg) {
        errorMessage = response.data.msg;
      } else if (response.data.errors) {
        const errorKeys = Object.keys(response.data.errors);
        if (errorKeys.length > 0) {
          errorMessage = response.data.errors[errorKeys[0]].msg;
        }
      }

      dispatch(onLogout(errorMessage));

      setTimeout(() => {
        dispatch(clearError());
      }, 5000);
    }
  };

  const startRegister = async ({ name, email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post('/auth/new', {
        name,
        email,
        password,
      });
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch ({ response }) {
      let errorMessage = 'Error en el servidor';

      if (response.data.msg) {
        errorMessage = response.data.msg;
      } else if (response.data.errors) {
        const errorKeys = Object.keys(response.data.errors);
        if (errorKeys.length > 0) {
          errorMessage = response.data.errors[errorKeys[0]].msg;
        }
      }

      dispatch(onLogout(errorMessage));

      setTimeout(() => {
        dispatch(clearError());
      }, 5000);
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token');
    if (!token) {
      dispatch(onLogout());
      return;
    }

    try {
      const { data } = await calendarApi.get('/auth/renew');
      localStorage.setItem('token', data.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  const startLogout = async () => {
    localStorage.clear();
    dispatch(onLogout());
  };

  return {
    //* Propiedades
    status,
    user,
    errorMessage,

    //* MÃ©todos
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
