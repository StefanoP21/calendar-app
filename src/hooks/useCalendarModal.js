import { useEffect, useState } from 'react';

import { addHours } from 'date-fns';
import { registerLocale } from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';

import { useCalendarStore, useUiStore } from './';

import es from 'date-fns/locale/es';

registerLocale('es', es);

const initialValues = {
  title: '',
  notes: '',
  start: new Date(),
  end: addHours(new Date(), 2),
  bgColor: '#0073ff',
  user: {},
};

export const useCalendarModal = () => {
  const { isDateModalOpen, openDateModal, closeDateModal } = useUiStore();
  const { activeEvent, setActiceEvent } = useCalendarStore();

  const [formValues, setFormValues] = useState(initialValues);
  const [formSubmitted, setFormSubmitted] = useState(false);

  useEffect(() => {
    if (activeEvent !== null) {
      setFormValues({ ...activeEvent });
    }
  }, [activeEvent]);

  const onInputChange = ({ target }) => {
    setFormValues({
      ...formValues,
      [target.name]: target.value,
    });
  };

  const onDateChange = (event, changing) => {
    setFormValues({
      ...formValues,
      [changing]: event,
    });
  };

  const handleNewEvent = () => {
    setActiceEvent(initialValues);

    openDateModal();
  };

  return {
    //* Propiedades
    isDateModalOpen,
    openDateModal,
    closeDateModal,
    formValues,
    formSubmitted,
    setFormSubmitted,

    //* MÃ©todos
    onInputChange,
    onDateChange,
    handleNewEvent,
  };
};
