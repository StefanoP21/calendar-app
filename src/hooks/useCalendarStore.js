import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useUiStore } from './useUiStore';
import {
  onAddNewEvent,
  onDeleteEvent,
  onSetActiveEvent,
  onUpdateEvent,
} from '../context';

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { openDateModal } = useUiStore();
  const [lastView, setLastView] = useState(
    localStorage.getItem('lastView') || 'month'
  );

  const eventStyleGetter = (event, start, end, isSelected) => {
    const style = {
      backgroundColor: event.bgColor,
      borderRadius: '0px',
      opacity: 0.8,
      color: 'white',
    };

    return {
      style,
    };
  };

  const onDoubleClick = (e) => {
    openDateModal();
  };

  const onViewChange = (e) => {
    localStorage.setItem('lastView', e);
    setLastView(e);
  };

  const setActiceEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      dispatch(onAddNewEvent({ _id: new Date().getTime(), ...calendarEvent }));
    }
  };

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  };

  return {
    //* Propiedades
    events,
    activeEvent,
    lastView,
    hasEventSelected: !!activeEvent,

    //* MÃ©todos
    eventStyleGetter,
    onDoubleClick,
    onViewChange,
    setActiceEvent,
    startSavingEvent,
    startDeletingEvent,
  };
};
