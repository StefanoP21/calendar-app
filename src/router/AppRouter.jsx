import { Navigate, Route, Routes } from 'react-router-dom';
import { LoginPage } from '../auth';
import { CalendarPage } from '../app';

export const AppRouter = () => {
  const authStatus = 'not-authenticated';

  return (
    <Routes>
      {authStatus === 'authenticated' ? (
        <Route path="/*" element={<CalendarPage />} />
      ) : (
        <Route path="/auth/*" element={<LoginPage />} />
      )}

      <Route path="/*" element={<Navigate to="/auth/login" />} />
    </Routes>
  );
};
