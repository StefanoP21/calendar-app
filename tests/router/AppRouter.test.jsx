import { MemoryRouter } from 'react-router-dom';
import { act, render, screen } from '@testing-library/react';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/app/pages/CalendarPage', () => ({
  CalendarPage: () => <div>CalendarRoutes</div>,
}));

describe('Pruebas en <AppRouter />', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('debe de mostrar la pantalla de carga y llamar a checkAuthToken', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'checking',
    });

    render(<AppRouter />);

    expect(screen.getByText('Verificando sesión...')).toBeTruthy();
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  test('debe de mostrar la pantalla de login', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'not-authenticated',
    });

    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(
      screen.getByRole('heading', { name: 'Iniciar sesión' })
    ).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar la pantalla de registro', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'not-authenticated',
    });

    const { container } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );
    const registerLink = screen.getByRole('link');

    act(() => {
      registerLink.click();
    });

    expect(screen.getByRole('heading', { name: 'Crear cuenta' })).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar el calendario si estamos auténticados', () => {
    useAuthStore.mockReturnValue({
      checkAuthToken: mockCheckAuthToken,
      status: 'authenticated',
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText('CalendarRoutes')).toBeTruthy();
  });
});
