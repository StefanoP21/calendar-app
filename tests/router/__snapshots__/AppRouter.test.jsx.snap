// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AppRouter /> debe de mostrar la pantalla de login 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="col-lg-8 login-form-1"
    >
      <h3>
        Iniciar sesión
      </h3>
      <form>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="email"
            placeholder="Correo electrónico"
            type="email"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="btnSubmit"
            type="submit"
            value="Ingresar"
          />
        </div>
      </form>
      <div
        class="form-group"
      >
        <a
          href="/auth/register"
        >
          ¿No tienes una cuenta?
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <AppRouter /> debe de mostrar la pantalla de registro 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="col-lg-8 login-form-1"
    >
      <h3>
        Crear cuenta
      </h3>
      <form>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="name"
            placeholder="Nombre de usuario"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="email"
            placeholder="Correo electrónico"
            type="email"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="form-control"
            name="confirmPassword"
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-group mb-3"
        >
          <input
            class="btnSubmit"
            type="submit"
            value="Registrarse"
          />
        </div>
      </form>
      <a
        href="/auth/login"
      >
        ¿Ya tienes una cuenta?
      </a>
    </div>
  </div>
</div>
`;
